endpoint: Fix goroutine leak when EP is deleted

christarazi opened the pull request:
The metadata resolver controller can be put in a scenario where it can
never succeed if the Cilium K8s pod cache is out-of-sync with the
current view of pods in the cluster. The cache can be out-of-sync if
there's heavy load on the cluster and especially on the apiserver. This
leaves the Cilium pod cache starved. When the pod cache is starved,
Cilium may never have an up-to-date view of the pods running in the
cluster, and therefore may never be able to resolve labels for pods
because the fetch will return "pod not found".

Eventually, kubelet (or maybe even the user) will give up on the pod and
remove it, thereby Cilium begins removing the endpoint. The controller
is stopped, but the goroutine waiting on the `done` channel will never
receive, hence the leak.

This commit fixes the goroutine leak when the controller never succeeds
and therefore, never closes the `done` channel. The fix is to add a
`select` statement to also watch the endpoint's `aliveCtx` which is
cancelled (and the context's `Done` channel is closed) when the endpoint
is deleted.

This commit was validated by forcing the metadata resolver to never find
a pod (manually hardcode the wrong pod name), and ensuring that the 
`gops stack` output does not contain an entry like:

```
goroutine 1244259 [chan receive, 1434 minutes]:
github.com/cilium/cilium/pkg/endpoint.(*Endpoint).RunMetadataResolver.func1(0xc0055d2a20, 0xc0049ff600, 0xc0055d2ae0, 0x5d)
	/go/src/github.com/cilium/cilium/pkg/endpoint/endpoint.go:1531 +0x34
created by github.com/cilium/cilium/pkg/endpoint.(*Endpoint).RunMetadataResolver
	/go/src/github.com/cilium/cilium/pkg/endpoint/endpoint.go:1530 +0x11e
```

Fixes: https://github.com/cilium/cilium/issues/13680

```release-note
Fix bug where Cilium leaks a goroutine when an endpoint is deleted. This leak, if left running in a high pod churn environment, can cause Cilium to exceed its memory usage and get OOM killed.
```

christarazi committed:
endpoint: Fix goroutine leak when EP is deleted

The metadata resolver controller can be put in a scenario where it can
never succeed if the Cilium K8s pod cache is out-of-sync with the
current view of pods in the cluster. The cache can be out-of-sync if
there's heavy load on the cluster and especially on the apiserver. This
leaves the Cilium pod cache starved. When the pod cache is starved,
Cilium may never have an up-to-date view of the pods running in the
cluster, and therefore may never be able to resolve labels for pods
because the fetch will return "pod not found".

Eventually, kubelet (or maybe even the user) will give up on the pod and
remove it, thereby Cilium begins removing the endpoint. The controller
is stopped, but the goroutine waiting on the `done` channel will never
receive, hence the leak.

This commit fixes the goroutine leak when the controller never succeeds
and therefore, never closes the `done` channel. The fix is to add a
`select` statement to also watch the endpoint's `aliveCtx` which is
cancelled (and the context's `Done` channel is closed) when the endpoint
is deleted.

This commit was validated by forcing the metadata resolver to never find
a pod (manually hardcode the wrong pod name), and ensuring that the
`gops stack` output does not contain an entry like:

```
goroutine 1244259 [chan receive, 1434 minutes]:
github.com/cilium/cilium/pkg/endpoint.(*Endpoint).RunMetadataResolver.func1(0xc0055d2a20, 0xc0049ff600, 0xc0055d2ae0, 0x5d)
	/go/src/github.com/cilium/cilium/pkg/endpoint/endpoint.go:1531 +0x34
created by github.com/cilium/cilium/pkg/endpoint.(*Endpoint).RunMetadataResolver
	/go/src/github.com/cilium/cilium/pkg/endpoint/endpoint.go:1530 +0x11e
```

Fixes: https://github.com/cilium/cilium/issues/13680

Signed-off-by: Chris Tarazi <chris@isovalent.com>

joestringer commented on pkg/endpoint/endpoint.go:
```diff
@@ -1642,7 +1642,11 @@ func (e *Endpoint) RunMetadataResolver(resolveMetadata MetadataResolverCB) {
-		<-done
+		select {
+		case <-done:
+		case <-e.aliveCtx.Done():
+			return
```
I suspect that another option would be to add a `StopFunc` to the controller below to close `done`, though that said I don't think there should be a functional difference so I'm happy with this solution.

christarazi replied:
Just for posterity, we could add a `StopFunc` to also close the `done` channel, but wouldn't work for the way we've set up this controller. Essentially, the configured run interval at 30s seems like it's there to provide a buffer for the initial run. Once it runs and succeeds, the controller closes `done` and returns, _and_ this would normally run again after 30s, but the goroutine is waiting for the `done` channel to be closed and removes the controller, thereby stopping it.

So having the close on `done` be inside `StopFunc` wouldn't work because we want to stop this controller as soon as it succeeds. 

I've prepared a commit to remove the interval because it is actually unnecessary. Waiting for CI to finish its run.

joestringer reviewed (approved):
LGTM. :shipit:

christarazi committed:
endpoint: Remove interval for metadata resolver

This controller's run interval is actually unnecessary as this
controller is intended to only run once as long as it succeeds. This is
evidenced by the fact that we have a `done` channel that's closed when
the controller succeeds and a goroutine waiting `done` to be closed to
remove the controller, thereby stopping it.

In addition, in the very unlikely case that the controller is scheduled
again to run _and_ the goroutine waiting for the `done` channel to be
closed has not run yet, the controller could panic because it would
close the channel twice. This commit prevents that as the interval
defaults to 10m if not provided. If the goroutine waiting on the channel
doesn't run within 10m, then we very likely have much worse problems at
hand.

Signed-off-by: Chris Tarazi <chris@isovalent.com>
